sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/RowLevelOperationBuilder.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.write;
+
+import org.apache.spark.annotation.Experimental;
+
+/**
+ * An interface for building a {@link RowLevelOperation}.
+ *
+ * @since 3.3.0
+ */
+@Experimental
+public interface RowLevelOperationBuilder {
+  /**
+   * Returns a {@link RowLevelOperation} that controls how Spark rewrites data
+   * for DELETE, UPDATE, MERGE commands.
+   */
+  RowLevelOperation build();
```

Comment:
I'm not against this API design but just want to put my idea: Shall we have different build methods for UPDATE, DELETE, MERGE? like `buildDelete`. Then we don't need `RowLevelOperationInfo`
```
public interface SupportsRowLevelOperations extends Table {
  RowLevelOperationBuilder newRowLevelOperationBuilder(CaseInsensitiveStringMap options);
}

public interface RowLevelOperationBuilder {
  RowLevelOperation buildDelete();
  RowLevelOperation buildUpdate();
  RowLevelOperation buildMerge();
}
```

Reply:
or we can keep the enum
```
public interface RowLevelOperationBuilder {
  RowLevelOperation build(Command command);
}
```

Reply:
Either way would work for me too.

I tried to match `LogicalWriteInfo`. One benefit of having an extra class is that we can extend it to pass more information in the future. Some of those future requirements are hard to predict when an API is designed. For instance, I plan to add a few new methods to `LogicalWriteInfo` for delta-based operations (e.g. row ID schema). However, there is no guarantee that will ever be required for `RowLevelOperationInfo`.

Any preference, @viirya @huaxingao @rdblue?

Reply:
Shall we stick to the current API just in case then?

Reply:
Sorry, my question was directed to @cloud-fan :)

Reply:
😄

Reply:
I know predicting the future is very hard, but I'm afraid we may over-design this. We have both `RowLevelOperationInfo` and the builder pattern to pass required information in the future, shall we keep only one of them?

Reply:
After a second thought, the read API doesn't have `ScanInformation` either. We pass options to create `ScanBuilder`, and then use different mix-it traits to pass scan information, and finally build the scan. The rationale is, we should pass the information that is always required to the builder, and the builder can use mix-in traits to take optional information.

That said, the current API design does follow the existing API style. +1 from me.